codegenerator/cli/templates/static/codegen/src/globalState/GlobalState.res

```diff
@@ -851,62 +826,85 @@ let executeNextQuery = (
   //Used for dependency injection for tests
   ~waitForNewBlock,
-  ~executeNextQuery,
+  ~executePartitionQuery,
   //required args
   ~state,
   ~dispatchAction,
 ) =>
   async chain => {
-    let chainFetcher = state.chainManager.chainFetchers->ChainMap.get(chain)
-    let {chainConfig: {chainWorker}, logger, currentBlockHeight} = chainFetcher
-
     if !isRollingBack(state) {
-      let (nextQuery, nextStateIfChangeRequired) =
-        chainFetcher->ChainFetcher.getNextQuery(~maxPerChainQueueSize=state.maxPerChainQueueSize)
+      let chainFetcher = state.chainManager.chainFetchers->ChainMap.get(chain)
+      let {chainConfig: {chainWorker}, logger, currentBlockHeight} = chainFetcher
 
-      switch nextStateIfChangeRequired {
-      | Some(nextFetchState) => dispatchAction(SetFetchState(chain, nextFetchState))
-      | None => ()
-      }
+      let (nextQueries, nextFetchState) =
+        chainFetcher->ChainFetcher.getNextQueries(~maxPerChainQueueSize=state.maxPerChainQueueSize)
+      dispatchAction(SetFetchState(chain, nextFetchState))
 
-      let setCurrentBlockHeight = currentBlockHeight =>
-        dispatchAction(SetFetchStateCurrentBlockHeight(chain, currentBlockHeight))
-
-      switch nextQuery {
-      | WaitForNewBlock =>
-        await waitForNewBlock(~logger, ~chainWorker, ~currentBlockHeight, ~setCurrentBlockHeight)
-      | NextQuery(queries) =>
-        let newPartitionsCurrentlyFetching =
-          queries->Array.map(query => query.partitionId)->Set.Int.fromArray
-        dispatchAction(SetCurrentlyFetchingBatch(chain, newPartitionsCurrentlyFetching))
-        let isPreRegisteringDynamicContracts =
-          state.chainManager->ChainManager.isPreRegisteringDynamicContracts
-        let _ =
-          await queries
-          ->Array.map(query =>
-            executeNextQuery(
+      let readyQueries = nextQueries->Js.Array2.filter(({fromBlock, toBlock}) =>
+        if fromBlock > currentBlockHeight {
+          false // This query should wait
+        } else {
+          switch toBlock {
+          | Some(toBlock) if toBlock < fromBlock =>
+            //This is an invalid case. We should never arrive at this match arm but it would be
+            //detrimental if it were the case.
+            FromBlockIsHigherThanToBlock({fromBlock, toBlock})->ErrorHandling.mkLogAndRaise(
               ~logger,
-              ~chainWorker,
-              ~currentBlockHeight,
-              ~setCurrentBlockHeight,
-              ~chain,
-              ~query,
-              ~dispatchAction,
-              ~isPreRegisteringDynamicContracts,
+              ~msg="Unexpected error getting next query in partition"
             )
-          )
-          ->Promise.all
+          | _ => ()
+          }
+          true
+        }
+      )
+
+      switch (readyQueries, nextQueries) {
+        | (_, []) => {
+          // This is an invalid case. Just do nothing as we did before.
+          // Maybe we should log this?
+          ()
```

Comment:
This is how it used to be before my change. Decided not to throw here, because I'm not sure that it won't happen